src/Logary.Prometheus/Exporter.fs

```diff
@@ -0,0 +1,171 @@
+open System
+open System.Text
+open System.Text.RegularExpressions
+open Logary.Metric
+open Suave
+open Suave.Filters
+open Suave.Operators
+open Suave.Writers
+open Hopac
+
+
+type ExporterConf =
+  {
+    metricNameTransformer: string -> string
+    metricRegistry: MetricRegistry
+    webConfig: SuaveConfig
+    urlPath: string
+  }
+
+module ExporterConf =
+  let validNameRegex = Regex("^[a-zA-Z_][a-zA-Z0-9_]*$", RegexOptions.Compiled)
+  let reservedNameRegex =  Regex("^__.*$", RegexOptions.Compiled)
+  let startCharValidRegex =  Regex("^[a-zA-Z_].*$", RegexOptions.Compiled)
+  let notValidAfterPartRegex =  Regex("[^a-zA-Z0-9_]", RegexOptions.Compiled)
+
+  let defaultMetricNameTrans metricName =
+    let mutable nameResult = metricName
+    if String.IsNullOrEmpty nameResult then failwith "name can't be null"
+    if reservedNameRegex.IsMatch nameResult then
+      nameResult <- "metric_" + nameResult.Substring 2
+    if not <| startCharValidRegex.IsMatch nameResult then
+      nameResult <- "metric_" + nameResult
+    notValidAfterPartRegex.Replace(nameResult, "_")
+
+  let nameValidator name =
+    if not <| validNameRegex.IsMatch name then
+      failwithf "name (%s) should match regex %s" name (validNameRegex.ToString())
+    elif reservedNameRegex.IsMatch name then
+      failwithf "(%s) is invalid, names beginning with __ are reserved for internal use." name
+    else name
+
+  let create (urlPath: string) (registry: MetricRegistry) =
+    {
+      metricNameTransformer = defaultMetricNameTrans
+      metricRegistry = registry
+      webConfig = defaultConfig
+      urlPath =  urlPath
+    }
+
+  let webConfig webConfig conf =
+    { conf with webConfig = webConfig }
+
+
+module Exporter =
+  open System.Threading
+  open System.Globalization
+  open Logary.Internals.Chiron.Formatting
+  open System.IO
+
+  /// HELP lines may contain any sequence of UTF-8 characters (after the metric name),
+  /// but the backslash and the line feed characters have to be escaped as \\ and \n, respectively
+  let escapeHelp (help: string) =
+    help.Replace("\\", @"\\").Replace("\n",@"\n")
+
+  /// label_value can be any sequence of UTF-8 characters,
+  /// but the backslash (\), double-quote ("), and line feed (\n) characters
+  /// have to be escaped as \\, \", and \n, respectively.
+  let escapeLabelValue (labelValue: string) =
+    labelValue.Replace("\\", @"\\").Replace("\"",@"\""").Replace("\n",@"\n")
+
+  let formatFloat (value: float) =
+    if Double.IsPositiveInfinity value then "+Inf"
+    elif Double.IsNegativeInfinity value then "-Inf"
+    elif Double.IsNaN value then "Nan"
+    else value.ToString(CultureInfo.InvariantCulture)
+
+  let appendLine (sb: StringBuilder) s =
+    sb.AppendLine s |> ignore
+
+  let getTypeStr (metricInfo: MetricInfo) =
+    match metricInfo with
+    | Gauge _ -> "gauge"
+    | Histogram _ -> "histogram"
+
+  let getIdentifier nameTransform (name: string) (labels: Map<string,string>) =
```

Comment:
Is it intended not to use Logary's `HashMap`? I know it's harder to use though. Same comment here as the formatting above about Span-based APIs - not required to fix though.

Reply:
labels are used in two cases, 
- when register a labeled metric, it will use labels(name-value paris) as the key of the metric, `let metricStore = new ConcurrentDictionary<Map<string,string>, 't>()`, `Map` can give the structural equality which `HashMap` doesn't.  no more lookups here.
- when export as metrc info to backend
  in this case, it will just loop for formatting, seems `HashMap` here give less advantage here.